MVU example

p69 opened the pull request:
Added example for MVU pattern using [dartea](https://pub.dartlang.org/packages/dartea).
All integration tests are passed, but I don't have mac, so I can't verify it on iOS. It would be very helpful if someone could try it on iOS.

brianegan commented on example/mvu/README.md:
```diff
@@ -0,0 +1,63 @@
+### Pros
+* #### Easy to understand and to maintain
+    All the app state is in one immutable object (Model), all the app logic is in one pure function (Update), all the view rendering logic is in one pure function (View)
+* #### Super easy to test
```
Could I please ask a bit more about testing and making sure I'm understanding the pattern correctly? If I follow this correctly, the update function returns an `Update`, which is a tuple of a model and effects.

However, when I looked at the tests, it seems like only half of the tuple is tested: The model. I couldn't spot any tests of the effects themselves. Are there strategies to test the Commands produced by the update function?

For example, in the update function of `/stats/state.dart`, the `LoadStats` branch of logic creates a `loadCmd`, which needs to call the repo and then dispatch another `OnStatsLoaded` Message. It feels like that logic is ripe for testing, first to make sure this command actually loads todos, and second to make sure you're doing right thing when the todos are loaded. 

What are your thoughts?

p69 replied:
Good question!
Yes, you are absolutely right - there are no tests for commands. I decided to exclude it from unit-tests, because I thought that integration tests should cover all the cases with commands. And, actually they covered it, but now I think that it's not right.

One way how to test commands is:
1. Create and provide commands factory to `update` function.
2. Implement this factory for tests.
3. Run command in test and verify that it triggers desired side-effect and generates correct `message`.

I'm going to try this approach today, seems like it should work.

brianegan commented on example/mvu/README.md:
```diff
@@ -0,0 +1,63 @@
+The heart of the MVU application are three yellow boxes on the diagram. First, the state of the app (Model) is mapped to the widgets tree (View). Second, events from the UI are translated into Messages and goes to the Update function (together with current app state). Update function is the brain of the app. It contains all the presentation logic, and it MUST be [pure](https://en.wikipedia.org/wi ...
+
+### Pros
+* #### Easy to understand and to maintain
```
Could we avoid the word easy in this case and think of slightly different wording? Just don't want someone to feel bad if they find this stuff difficult to understand, as this stuff can be for newer folks.

p69 commented on example/mvu/lib/home/view.dart:
```diff
@@ -0,0 +1,71 @@
+        (m) => dispatch(new TodosMsg(m)), body.model);
+    var filterMenu = todos.buildFilterMenu(
+        context, (m) => dispatch(new TodosMsg(m)), body.model);
+    return [extraActions, filterMenu];
```
Fixed, thank you.

p69 commented:
@brianegan Thank you for such a quick review!
I'm going to add some examples of commands testing.

p69 commented:
Added some tests for `effects` part in result of `update` function. Also added tests for side-effects.
@brianegan please take a look again, thanks.

brianegan commented:
Will do, thanks! Might be this evening or tomorrow :)

brianegan commented:
Hey there -- Sorry about the delay! A bit busy with travel this week. The updates look really good to me. Thanks for providing examples of how to test the Command logic as well so we folks can evaluate that part of the pattern.

Again, thanks so much and amazing work overall 😄 🎉 🎈 👏
